internal/appsec/dyngo/operation.go

```diff
@@ -141,107 +129,103 @@ func NewRootOperation() Operation {
-func (o *operation) Parent() Operation {
-	return o.parent
+	return &operation{parent: parent.unwrap()}
 }
 
-// FinishOperation finishes the operation along with its results and emits a
+// Finish finishes the operation along with its results and emits a
 // finish event with the operation results.
 // The operation is then disabled and its event listeners removed.
-func FinishOperation(op Operation, results interface{}) {
-	op.finish(op, results)
-}
+func Finish[O Operation, E ResultOf[O]](op O, results E) {
+	o := op.unwrap()
+	defer o.disable() // This will need the RLock below to be released...
 
-func (o *operation) finish(op Operation, results interface{}) {
-	// Defer the call to o.disable() first so that the RWMutex gets unlocked first
-	defer o.disable()
 	o.mu.RLock()
 	defer o.mu.RUnlock() // Deferred and stacked on top of the previously deferred call to o.disable()
+
 	if o.disabled {
 		return
 	}
-	resType := reflect.TypeOf(results)
-	for current := op; current != nil; current = current.Parent() {
-		current.emitEvent(resType, op, results)
+
+	for current := o; current != nil; current = current.parent {
+		emitEvent(&current.eventRegister, op, results)
 	}
 }
 
 // Disable the operation and remove all its event listeners.
 func (o *operation) disable() {
 	o.mu.Lock()
 	defer o.mu.Unlock()
+
 	if o.disabled {
 		return
 	}
+
 	o.disabled = true
 	o.eventRegister.clear()
 }
 
-// Add the given event listeners to the operation.
-func (o *operation) add(l ...EventListener) {
+// On registers and event listener that will be called when the operation
+// begins.
+func On[O Operation, E ArgOf[O]](op Operation, l EventListener[O, E]) {
+	o := op.unwrap()
+
 	o.mu.RLock()
 	defer o.mu.RUnlock()
 	if o.disabled {
 		return
 	}
-	for _, l := range l {
-		if l == nil {
-			continue
-		}
-		key := l.ListenedType()
-		o.eventRegister.add(key, l)
-	}
+	addEventListener(&o.eventRegister, l)
 }
 
-// On registers the event listener. The difference with the Register() is that
-// it doesn't return a function closure, which avoids unnecessary allocations
-// For example:
-//
-//	op.On(MyOperationStart(func (op MyOperation, args MyOperationArgs) {
-//	    // ...
-//	}))
-func (o *operation) On(l EventListener) {
+// OnFinish registers an event listener that will be called when the operation
+// finishes.
+func OnFinish[O Operation, E ResultOf[O]](op Operation, l EventListener[O, E]) {
+	o := op.unwrap()
+
 	o.mu.RLock()
 	defer o.mu.RUnlock()
 	if o.disabled {
 		return
 	}
-	o.eventRegister.add(l.ListenedType(), l)
+	addEventListener(&o.eventRegister, l)
 }
 
-func (o *operation) OnData(l DataListener) {
+func OnData[T any](op Operation, l DataListener[T]) {
+	o := op.unwrap()
+
 	o.mu.RLock()
 	defer o.mu.RUnlock()
 	if o.disabled {
 		return
 	}
-	o.dataBroadcaster.add(l.ListenedType(), l)
+	addDataListener(&o.dataBroadcaster, l)
 }
 
-func (o *operation) EmitData(data any) {
+func EmitData[T any](op Operation, data T) {
```

Comment:
Nit: description to keep exported functions documented.
Sorry that one is on me 🙈